### PR TITLE
Unsupervised pre-training of recognizer

### PR DESCRIPTION
**Pull request by mittagessen**
This is an implementation for unsupervised pretraining of recognition model weights based on an image inpainting surrogate task that aims to
reconstruct randomly sampled masked patches from the initial convolutional
feature maps that have been replaced with a learnable embedding. The model is
trained with a contrastive loss where negative samples are randomly generated
from the unmasked parts of the sequence.

Apart from the sampling method the implementation is mostly a faithful
adaptation of:

```
Vogler, Nikolai, et al. "Lacuna Reconstruction: Self-supervised Pre-training
for Low-Resource Historical Document Transcription." arXiv preprint
arXiv:2112.08692 (2021).
```

Status of the PR
-------------------------
- [X] Dataset support
- [X] Wav2Vec2 masking layer
- [X] Pytorch-lightning model
- [X] CLI driver
- [X] Auxiliary layers in VGSL model
- [x] Deserialization of auxiliary layers
- [x] Docs

### CODE REVIEW

File: kraken/lib/pretrain/layers.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        super().__init__()
+
+        self.context_encoder_input_dim = context_encoder_input_dim
+        self.final_dim = final_dim
+        self.mask_width = mask_width
+        self.mask_prob = mask_prob
+        self.num_negatives = num_negatives
+
+        # mask embedding replacing the masked out areas
+        self.mask_emb = Embedding(mask_width, context_encoder_input_dim)
```
Comment (nvog) [inline kraken/lib/pretrain/layers.py:56]: it's maybe worth having a _separate_ knob we can set for width of the mask embedding itself, i.e., so that we can try using either a `mask_width` embedding here or a 1 time-step wide mask that gets repeated like original paper

Comment (mittagessen) [reply kraken/lib/pretrain/layers.py:56]: I'm loath to add switches where it should be possible to find an 'optimal' value. I'd evaluate both this formulation and the original one and see which one performs better and then just use that one.
